app/attributes/attributecontroller.cpp

```diff
@@ -941,6 +941,78 @@ bool AttributeController::save()
   if ( !mFeatureLayerPair.layer() )
     return false;
 
+  const QStringList photoFields = QgsProject::instance()->entryList( QStringLiteral( "Mergin" ), QStringLiteral( "PhotoNaming/%1" ).arg( mFeatureLayerPair.layer()->id() ) );
+
+  QgsExpressionContext expressionContext = mFeatureLayerPair.layer()->createExpressionContext();
+  expressionContext << QgsExpressionContextUtils::formScope( mFeatureLayerPair.feature() );
+  if ( mVariablesManager )
+    expressionContext << mVariablesManager->positionScope();
+
+  expressionContext.setFields( mFeatureLayerPair.feature().fields() );
+  expressionContext.setFeature( mFeatureLayerPair.featureRef() );
+
+  // check for new photos
+  QMap<QUuid, std::shared_ptr<FormItem>>::iterator formItemsIterator = mFormItems.begin();
+  while ( formItemsIterator != mFormItems.end() )
+  {
+    std::shared_ptr<FormItem> item = formItemsIterator.value();
+    if ( item->type() == FormItem::Field && item->editorWidgetType() == QStringLiteral( "ExternalResource" ) )
+    {
+      QVariantMap config = item->editorWidgetConfig();
+      if ( config[ QStringLiteral( "DocumentViewer" ) ] != 1 )
+      {
+        continue;
+      }
+
+      const QgsField field = item->field();
+      if ( item->originalValue() != mFeatureLayerPair.feature().attribute( item->fieldIndex() ) && photoFields.contains( field.name() ) )
+      {
+        QString expString = QgsProject::instance()->readEntry( QStringLiteral( "Mergin" ), QStringLiteral( "PhotoNaming/%1/%2" ).arg( mFeatureLayerPair.layer()->id() ).arg( field.name() ) );
+        QgsExpression exp( expString );
+        exp.prepare( &expressionContext );
+        if ( exp.hasParserError() )
+        {
+          QgsMessageLog::logMessage( tr( "Expression for %1:%2 has parser error: %3" ).arg( mFeatureLayerPair.layer()->name(), field.name(), exp.parserErrorString() ), QStringLiteral( "Input" ), Qgis::Warning );
+          continue;
+        }
+
+        QVariant value = exp.evaluate( &expressionContext );
+        if ( exp.hasEvalError() )
+        {
+          QgsMessageLog::logMessage( tr( "Expression for %1:%2 has evaluation error: %3" ).arg( mFeatureLayerPair.layer()->name(), field.name(), exp.evalErrorString() ), QStringLiteral( "Input" ), Qgis::Warning );
+        }
+        else
+        {
+          QVariant val( value );
+          if ( !field.convertCompatible( val ) )
+          {
+            QgsMessageLog::logMessage( tr( "Value \"%1\" %4 could not be converted to a compatible value for field %2 (%3)." ).arg( value.toString(), field.name(), field.typeName(), value.isNull() ? "NULL" : "NOT NULL" ), QStringLiteral( "Input" ), Qgis::Warning );
+          }
+          else
+          {
```

Comment:
I would again prefer to `++formItemsIterator;` and `continue;` instead of nesting deeper